### PR TITLE
Remove IVTs to liveshare

### LINKED ISSUE #35074: Migrate Live Share IVTs to the external access pattern
We are planning to remove IVT access to dotnet/roslyn from the following assemblies for 16.2 Preview 1:

* Microsoft.VisualStudio.LanguageServices.Remote
* Microsoft.VisualStudio.LanguageServices.Remote.15.7
* Microsoft.VisualStudio.LanguageServices.Remote.15.8
* Microsoft.VisualStudio.LanguageServices.Remote.16.0
* Microsoft.VisualStudio.LanguageServices.Remote.16.1
* Microsoft.VisualStudio.LanguageServices.Remote.CSharp
* Microsoft.VisualStudio.LanguageServices.Remote..CSharp.15.7
* Microsoft.VisualStudio.LanguageServices.Remote.CSharp.15.8
* Microsoft.VisualStudio.LanguageServices.Remote.CSharp.16.0
* Microsoft.VisualStudio.LanguageServices.Remote.CSharp.16.1

These assemblies should be updated to operate correctly without IVT access to roslyn, or use an External Access assembly for all access to Roslyn internals. I will be assisting in the creation of the External Access layer if necessary for these components.

### PR DESCRIPTION
**Pull request by dibarbet**
Tested locally with liveshare running.  Runs fine.
Liveshare build should not be affected as the older DLLs build with specific roslyn versions.

Closes #35074

### CODE REVIEW

**Comment by heejaechang** [comment]
this is not in hurry right? let's check this in once I checked in the remote syntactic classifier thing.

**Comment by dibarbet** [comment]
> this is not in hurry right? let's check this in once I checked in the remote syntactic classifier thing.

yep!  can wait until that is in

File: src/VisualStudio/Setup/source.extension.vsixmanifest [hunk 1]
```diff
@@ -57,21 +57,21 @@
...
     <Asset Type="Microsoft.VisualStudio.MefComponent" Path="Microsoft.CodeAnalysis.ExternalAccess.FSharp.dll" />
     <Asset Type="Microsoft.VisualStudio.MefComponent" Path="Microsoft.CodeAnalysis.ExternalAccess.IntelliTrace.dll" />
-    <Asset Type="Microsoft.VisualStudio.MefComponent" Path="Microsoft.CodeAnalysis.ExternalAccess.LiveShare.dll" />
     <Asset Type="Microsoft.VisualStudio.MefComponent" Path="Microsoft.CodeAnalysis.ExternalAccess.ProjectSystem.dll" />
     <Asset Type="Microsoft.VisualStudio.MefComponent" Path="Microsoft.CodeAnalysis.ExternalAccess.TypeScript.dll" />
     <Asset Type="Microsoft.VisualStudio.MefComponent" Path="Microsoft.CodeAnalysis.ExternalAccess.UnitTesting.dll" />
     <Asset Type="Microsoft.VisualStudio.MefComponent" Path="Microsoft.CodeAnalysis.LanguageServer.Protocol.dll" />
     <Asset Type="Microsoft.VisualStudio.Analyzer" Path="Microsoft.CodeAnalysis.ExternalAccess.FSharp.dll" />
     <Asset Type="Microsoft.VisualStudio.MefComponent" Path="Microsoft.VisualStudio.LanguageServices.LiveShare.dll" />
+    <Asset Type="Microsoft.VisualStudio.MefComponent" Path="Microsoft.VisualStudio.LanguageServices.LiveShare.dll" />
```
Comment (sharwell) [inline src/VisualStudio/Setup/source.extension.vsixmanifest]: 📝 Duplicate line here

File: src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj [hunk 2]
```diff
@@ -83,20 +83,9 @@
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.Test.Utilities2" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.FSharp" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.FSharp.UnitTests" />
-    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.LiveShare" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.TypeScript.EditorFeatures" Key="$(TypeScriptKey)" WorkItem="https://github.com/dotnet/roslyn/issues/35077" />
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.TypeScript" Key="$(TypeScriptKey)" WorkItem="https://github.com/dotnet/roslyn/issues/35077" />
     <InternalsVisibleTo Include="Roslyn.Services.Editor.TypeScript.UnitTests" Key="$(TypeScriptKey)" WorkItem="https://github.com/dotnet/roslyn/issues/35077" />
-    <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.Remote" Key="$(RemoteLanguageServiceKey)" WorkItem="https://github.com/dotnet/roslyn/issues/35074" />
```
Comment (sharwell) [inline src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj]: ❔  Are all items linked to #35074 now resolved? If so, this pull request should close that issue.

Comment (dibarbet) [reply src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj]: Yep, updated the description.

**Comment by dibarbet** [comment]
@sharwell @jasonmalinowski would you be able to take a look?  would like to get all the moving stuff around in before I make more changes to this area

File: src/Features/LanguageServer/Protocol/Microsoft.CodeAnalysis.LanguageServer.Protocol.csproj [hunk 3]
```diff
@@ -24,7 +24,6 @@
   </ItemGroup>
 
   <ItemGroup>
-    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.LiveShare" />
```
Comment (genlu) [inline src/Features/LanguageServer/Protocol/Microsoft.CodeAnalysis.LanguageServer.Protocol.csproj]: So this basically is the new external access assembly for live share?

**Comment by genlu** [comment]
> namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare.Tagger

Do we need to change namespace for this one too?

---
Refers to: src/VisualStudio/LiveShare/Impl/Client/Tagger/TextMateClassificationTaggerProvider.cs:11 in cdbf3c4. [](commit_id = cdbf3c48a4c4713c2471c2559d4eccccc69c107a, deletion_comment = False)

File: src/VisualStudio/LiveShare/Impl/Client/LocalForwarders/VBRemoteCompilationFactoryService.cs [hunk 4]
```diff
@@ -3,6 +3,7 @@
 using System.Composition;
 using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.LanguageServices.LiveShare.Client;
 
 namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare.LocalForwarders
```
Comment (genlu) [inline src/VisualStudio/LiveShare/Impl/Client/LocalForwarders/VBRemoteCompilationFactoryService.cs]: Do we need to change namespace for this one too?

File: src/Features/LanguageServer/Protocol/Microsoft.CodeAnalysis.LanguageServer.Protocol.csproj (hunk 3 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Microsoft.CodeAnalysis.LanguageServer.Protocol.csproj]: Kind of.  This contains all the stuff needed for liveshare integration.  But Liveshare does not have IVTs to this or any other assembly anymore.

File: src/VisualStudio/LiveShare/Impl/Client/LocalForwarders/CSharpRemoteCompilationFactoryService.cs [hunk 5]
```diff
@@ -3,6 +3,7 @@
 using System.Composition;
 using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.LanguageServices.LiveShare.Client;
 
 namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare
```
Comment (genlu) [inline src/VisualStudio/LiveShare/Impl/Client/LocalForwarders/CSharpRemoteCompilationFactoryService.cs]: Do we need to change namespace for this one too?

**Comment by genlu** [comment]
> namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare.Tagger

It seems all namespaces under Tagger folder remain unchanged, is it intentional?

---
Refers to: src/VisualStudio/LiveShare/Impl/Client/Tagger/AbstractAsyncClassificationTaggerProvider.cs:15 in cdbf3c4. [](commit_id = cdbf3c48a4c4713c2471c2559d4eccccc69c107a, deletion_comment = False)

**Comment by dibarbet** [comment]
> 
> 
> > namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare.Tagger
> 
> It seems all namespaces under Tagger folder remain unchanged, is it intentional?
> 
> Refers to: src/VisualStudio/LiveShare/Impl/Client/Tagger/AbstractAsyncClassificationTaggerProvider.cs:15 in [cdbf3c4](https://github.com/dotnet/roslyn/commit/cdbf3c48a4c4713c2471c2559d4eccccc69c107a). [](commit_id = [cdbf3c4](https://github.com/dotnet/roslyn/commit/cdbf3c48a4c4713c2471c2559d4eccccc69c107a), deletion_comment = False)

Nope, just didn't catch these new files in the merge.  Updating.

**Pull request merged**